src/main/java/com/github/jbellis/jvector/graph/GraphSearcher.java

```diff
@@ -92,17 +92,49 @@ public GraphSearcher build() {
   }
 
   public NeighborQueue search(
-      NeighborSimilarity.ScoreFunction scoreFunction,
+      NeighborSimilarity.ExactScoreFunction scoreFunction,
       int topK,
       Bits acceptOrds,
-      int visitedLimit) {
+      int visitedLimit)
+  {
     int initialEp = view.entryNode();
     if (initialEp == -1) {
       return new NeighborQueue(1, false);
     }
     NeighborQueue results;
     results = new NeighborQueue(topK, false);
     searchInternal(scoreFunction, results, topK, view.entryNode(), acceptOrds, visitedLimit);
+
+    return results;
+  }
+
+  public NeighborQueue search(
+          NeighborSimilarity.ExactScoreFunction exactScoreFunction,
+          NeighborSimilarity.ApproximateScoreFunction approximateScoreFunction,
+          int topK,
+          Bits acceptOrds,
+          int visitedLimit)
+  {
+    int initialEp = view.entryNode();
+    if (initialEp == -1) {
+      return new NeighborQueue(1, false);
+    }
+    NeighborQueue results;
+    results = new NeighborQueue(topK, false);
+    searchInternal(approximateScoreFunction, results, topK, view.entryNode(), acceptOrds, visitedLimit);
+
+    // re-rank the quantized results by the original similarity function
+    // Decorate
+    // TODO: if we capture ef/actual topK here, we can avoid rerank when nn.size < topK
```

Comment:
If you mean, because the definition of recall doesn't care where it is in the results, that's true, but I think it would surprise users not to get them back in actual similarity order